web/src/copy_and_paste.ts

```diff
@@ -199,8 +231,47 @@ export async function copy_handler(): Promise<void> {
     // * Otherwise, we want to copy the bodies of all messages that
     //   were partially covered by the selection.
 
+    // This is done to make the copying within math blocks smarter.
+    // We mutate the selection for selecting singular katex expressions
+    // within math blocks when applicable, which will be
+    // converted into the inline $$<expr>$$ syntax.
+
+    // In case when a single expression or its subset is selected
+    // within a math block, we adjust the selection so that it
+    // selects the katex span which is parenting that expression.
+
+    // We do not want to do this if the selection
+    // spreads across multiple katex displays or both the
+    // focus and anchor are not within the same katex-display.
+    // That is why we call `undo_katex_selection_if_applicable`
+    // at the end.
     const selection = window.getSelection();
     assert(selection !== null);
+    const original_range = selection.getRangeAt(0);
+
+    const anchor_element = get_closest_element(selection.anchorNode);
+    const focus_element = get_closest_element(selection.focusNode);
+    if (anchor_element) {
+        const parent = anchor_element.closest(".katex-display");
+        const is_math_block = parent !== null && parent !== selection.anchorNode;
+        if (is_math_block) {
+            const range = document.createRange();
+            range.selectNodeContents(parent);
+            selection.removeAllRanges();
+            selection.addRange(range);
+        }
+    } else if (focus_element) {
+        const parent = focus_element.closest(".katex-display");
+        const is_math_block = parent !== null && parent !== selection.focusNode;
+        if (is_math_block) {
+            const range = document.createRange();
+            range.selectNodeContents(parent);
+            selection.removeAllRanges();
+            selection.addRange(range);
+        }
+    }
+    undo_katex_selection_if_applicable(focus_element, anchor_element, selection, original_range);
```

Comment:
What I'd had in mine is the entire section of this function be `improve_katex_selection_range`, rather than just the second block. Sorry if that was unclear.